include/odp/api/spec/packet_io_types.h

```diff
@@ -1114,6 +1120,20 @@ typedef struct odp_lso_profile_param_t {
 			/** Field size in bytes. Valid values are 1, 2, 4, and 8 bytes. */
 			uint8_t size;
 
+			/** Operation specific parameters */
+			union {
+				/** parameters for ODP_LSO_ADD_SHIFTED_PAYLOAD_OFFSET */
+				struct {
+					/** Number of bits to shift the number of payload bytes
+					 *  right before adding it to the value in packet.
+					 *  Affects also segmentation so that segment boundaries
+					 *  are at offsets divisible by 1<<shift (i.e. the
+					 *  least significant bits shifted away are always zero).
+					 */
```

Comment:
The shift was applied to the payload_offset/fragment_offset as the name suggests, not to the payload bytes. The payload length was the same for all segments except the last one (based on total packet length), and it is divisible by (1 << shift). The payload_offset was the cumulative length of the previous fragment’s payload lengths. The shift was applied to this cumulative payload_offset to fit it into the header field. With the shift, the Nth fragment’s payload offset should be:

[ ((n - 1) * payload_length) >> shift ]

Reply:
Maybe I should read "was" as "should be" in the comment? Since this is a proposed new feature in the API, there are no pre-existing implementations of it. That said, if something in the proposal is somehow inconsistent or hard to implement (e.g. due to potential HW limitations), you can point out the issues and propose changes.

Since the shifting causes the least significant bits of the segment offset to not be present in the packet, it would be problematic if they would be nonzero (i.e. if the segment offset were not divisible by 1<<shift). How would the receiver then know what the true offset is? Because of that, this API change proposal specifies that the segmentation occurs at suitable boundaries for the shift value. Now I notice that the API text should be improved/fixed by changing "are at offsets" to "are at payload offsets", but other than that, I do not see what the problem is.

> The shift was applied to the payload_offset/fragment_offset as the name suggests, not to the payload bytes.

I am not sure I understand this comment. According to the API text, the shift is not applied to payload bytes, but the _number_ of payload bytes (in all previous segments). And the number of payload bytes in all previous segments is the same as the payload offset.

> The payload length was the same for all segments except the last one (based on total packet length), and it is divisible by (1 << shift). 

According to the existing API spec, the payload length is not necessarily the same in all segments of a single packet and there is no guarantee that the payload length is divisible by anything. See odp_packet_lso_opt_t::max_payload_len. Therefore I propose to add this statement in the API, to guarantee that segment boundaries are suitable if the offset shifting is used.

> The payload_offset was the cumulative length of the previous fragment’s payload lengths. The shift was applied to this cumulative payload_offset to fit it into the header field 

Yes, isn't that what the API says?

> With the shift, the Nth fragment’s payload offset should be: [ ((n - 1) * payload_length) >> shift ]

Only if the underlying implementation works so that all the previous segments have payload_length payload length.

Reply:
>2) separate the payload offset divisibility requirement from the shift value to a separate parameter.
That should be fine. Please provide a proposal that ensures if we select a non-zero shift value, then payload_length should be divisible by (1 << shift).

Reply:
As discussed offline, max_payload_len of odp_packet_lso_opt_t specifies the maximum payload length in a segment but allows ODP implementations to create segments with less payload bytes, e.g. due to HW limitations or for alignment purposes. Thus having the shift value affect the generated segments lengths should be ok. We can still consider having a separate parameter, but maybe it is not really that useful.